File: windows/Marquee.Core.Tests/RequestLimitsTests.cs
using Marquee.Core.Models;
using Marquee.Core.Tests.Support;

namespace Marquee.Core.Tests;

// Request limits and the trusted role (0.39+): /me's requestLimits and the
// member's line on the Requests page (app/requests/page.tsx quotaLine), the
// household member's quota fields, the admin's Role / Request limits form
// as PATCH /users/{id} sends it, who reviews requests (lib/users/roles.ts),
// and an older server's answers without any of it.

public sealed class RequestLimitsTests
{
    private static readonly TimeZoneInfo Utc = TimeZoneInfo.Utc;

    private static RequestLimit Limit(int limit, int days, int used, DateTimeOffset? next = null) => new()
    {
        Limit = limit,
        Days = days,
        Used = used,
        Remaining = Math.Max(0, limit - used),
        NextSlotAt = next,
    };

    // MARK: /me

    [Fact]
    public void MeFixtureHasNoLimitsForTheAdmin()
    {
        var me = Fixtures.Decode<Me>("me");

        Assert.NotNull(me.RequestLimits);
        Assert.Null(me.RequestLimits!.Movie);
        Assert.Null(me.RequestLimits.Tv);
        Assert.Null(me.RequestLimits.Summary(Utc));
        Assert.True(me.ReviewsRequests);
    }

    [Fact]
    public void MeDecodesALimit()
    {
        var json = Fixtures.Read("me")
            .Replace("\"role\": \"admin\"", "\"role\": \"member\"", StringComparison.Ordinal)
            .Replace("\"movie\": null", """
                "movie": { "limit": 5, "days": 7, "used": 5, "remaining": 0, "nextSlotAt": "2026-10-03T02:53:36.305Z" }
                """, StringComparison.Ordinal);
        var me = Json.Decode<Me>(json);

        var movie = Assert.IsType<RequestLimit>(me.RequestLimits?.Movie);
        Assert.Equal(5, movie.Limit);
        Assert.Equal(7, movie.Days);
        Assert.Equal(5, movie.Used);
        Assert.Equal(0, movie.Remaining);
        Assert.Equal(Json.ParseDate("2026-10-03T02:53:36.305Z"), movie.NextSlotAt);
        Assert.Null(me.RequestLimits!.Tv);
        Assert.False(me.ReviewsRequests);
        Assert.Equal("Movies: none left until Oct 3", me.RequestLimits.Summary(Utc));
    }

    [Fact]
    public void OlderServerMeHasNoLimits()
    {
        // Windows checkouts can turn the fixture's line endings into \r\n.
        var json = Fixtures.Read("me").Replace("\r\n", "\n", StringComparison.Ordinal);
        var start = json.IndexOf(",\n  \"requestLimits\"", StringComparison.Ordinal);
        Assert.True(start > 0);
        var older = json[..start] + "\n}";
        var me = Json.Decode<Me>(older);

        Assert.Null(me.RequestLimits);
        Assert.Equal("Timmy", me.Label);
    }

    // MARK: The member's line

    [Fact]
    public void LineWithRequestsLeft() =>
        Assert.Equal("Movies: 3 of 5 requests left (every 7 days)", Limit(5, 7, 2).Line("Movies", Utc));

    [Fact]
    public void LineWithNoneLeft() =>
        Assert.Equal("TV: none left until Oct 3",
            Limit(5, 7, 5, Json.ParseDate("2026-10-03T02:53:36.305Z")).Line("TV", Utc));

    [Fact]
    public void LineWithNoneLeftAndNoDate() =>
        Assert.Equal("Movies: none left", Limit(5, 7, 5).Line("Movies", Utc));

    [Fact]
    public void LineDateIsLocal()
    {
        var west = TimeZoneInfo.CreateCustomTimeZone("west", TimeSpan.FromHours(-5), "west", "west");
        Assert.Equal("Movies: none left until Oct 2",
            Limit(5, 7, 5, Json.ParseDate("2026-10-03T02:53:36.305Z")).Line("Movies", west));
    }

    [Fact]
    public void SummaryJoinsBothTypes()
    {
        var limits = new RequestLimits
        {
            Movie = Limit(5, 7, 2),
            Tv = Limit(2, 30, 2, Json.ParseDate("2026-10-03T12:00:00Z")),
        };
        Assert.Equal("Movies: 3 of 5 requests left (every 7 days) · TV: none left until Oct 3", limits.Summary(Utc));
        Assert.Equal("TV: none left until Oct 3", (limits with { Movie = null }).Summary(Utc));
    }

    // MARK: Roles

    [Fact]
    public void WhoReviewsRequests()
    {
        Assert.True(UserRole.Admin.ReviewsRequests);
        Assert.True(UserRole.Trusted.ReviewsRequests);
        Assert.False(UserRole.Member.ReviewsRequests);
        // An unknown role acts as a member.
        Assert.False(UserRole.FromValue("guest").ReviewsRequests);
        Assert.True(UserRole.Trusted.IsKnown);
        Assert.Equal("Trusted", UserRole.Trusted.Label);
    }

    [Fact]
    public void TrustedUserReviewsButIsNotAdmin()
    {
        var json = Fixtures.Read("me").Replace("\"role\": \"admin\"", "\"role\": \"trusted\"", StringComparison.Ordinal);
        var user = Json.Decode<Me>(json).User;

        Assert.Equal(UserRole.Trusted, user.Role);
        Assert.True(user.ReviewsRequests);
        Assert.False(user.IsAdmin);
    }

    // MARK: Household member

    [Fact]
    public void HouseholdMemberDecodesQuotas()
    {
        var member = Fixtures.Decode<HouseholdMember>("household-member");

        Assert.Equal(5, member.MovieQuotaLimit);
        Assert.Equal(7, member.MovieQuotaDays);
        Assert.Null(member.TvQuotaLimit);
        Assert.Equal(7, member.TvQuotaDays);
        Assert.True(member.SupportsRequestLimits);
        Assert.False(member.IsTrusted);
        Assert.Equal(UserRole.Member, MemberAccessForm.InitialRole(member));
        Assert.Equal("5", MemberAccessForm.LimitText(member.MovieQuotaLimit));
        Assert.Equal("", MemberAccessForm.LimitText(member.TvQuotaLimit));
        Assert.Equal("7", MemberAccessForm.DaysText(member.TvQuotaDays));
    }

    [Fact]
    public void TrustedMemberDecodes()
    {
        var json = Fixtures.Read("household-member").Replace("\"role\": \"member\"", "\"role\": \"trusted\"", StringComparison.Ordinal);
        var member = Json.Decode<HouseholdMember>(json);

        Assert.True(member.IsTrusted);
        Assert.False(member.IsAdmin);
        Assert.Equal(UserRole.Trusted, MemberAccessForm.InitialRole(member));
    }

    [Fact]
    public void OlderServerMemberHasNoQuotaFields()
    {
        var json = Fixtures.Read("household-member").Replace("\r\n", "\n", StringComparison.Ordinal);
        var start = json.IndexOf(",\n  \"movieQuotaLimit\"", StringComparison.Ordinal);
        Assert.True(start > 0);
        var member = Json.Decode<HouseholdMember>(json[..start] + "\n}");

        Assert.Null(member.MovieQuotaLimit);
        Assert.Null(member.MovieQuotaDays);
        Assert.False(member.SupportsRequestLimits);
        Assert.Equal("7", MemberAccessForm.DaysText(member.MovieQuotaDays));
    }

    [Fact]
    public void UpdateResultCarriesQuotas()
    {
        var update = Fixtures.Decode<UpdateUserResult>("user-update");
        Assert.Equal(5, update.User.MovieQuotaLimit);
        Assert.Null(update.User.TvQuotaLimit);
    }

    // MARK: PATCH body

    [Fact]
    public void AccessFormSendsRoleAndLimits()
    {
        var (request, error) = MemberAccessForm.Apply(new UpdateUserRequest("kid"), trusted: true, " 5 ", "7", "", "30");

        Assert.Null(error);
        Assert.Equal(
            """{"username":"kid","role":"trusted","movieQuotaLimit":5,"movieQuotaDays":7,"tvQuotaLimit":null,"tvQuotaDays":30}""",
            Json.EncodeBodyToString(request!));
    }

    [Fact]
    public void AccessFormBlankDaysAreLeftOut()
    {
        var (request, error) = MemberAccessForm.Apply(new UpdateUserRequest("kid"), trusted: false, "", "", "", " ");

        Assert.Null(error);
        Assert.Equal(
            """{"username":"kid","role":"member","movieQuotaLimit":null,"tvQuotaLimit":null}""",
            Json.EncodeBodyToString(request!));
    }

    [Theory]
    [InlineData("0", "7", "", "7", "The movie limit is a number from 1 to 1000, or blank for no limit.")]
    [InlineData("1001", "7", "", "7", "The movie limit is a number from 1 to 1000, or blank for no limit.")]
    [InlineData("", "7", "two", "7", "The TV limit is a number from 1 to 1000, or blank for no limit.")]
    [InlineData("5", "366", "", "7", "The number of days is from 1 to 365.")]
    [InlineData("5", "7", "", "-1", "The number of days is from 1 to 365.")]
    public void AccessFormRefusesBadValues(string movieLimit, string movieDays, string tvLimit, string tvDays, string message)
    {
        var (request, error) = MemberAccessForm.Apply(new UpdateUserRequest("kid"), false, movieLimit, movieDays, tvLimit, tvDays);

        Assert.Null(request);
        Assert.Equal(message, error);
    }

    [Fact]
    public void UpdateBodyWithoutAccessFieldsIsUnchanged() =>
        Assert.Equal("""{"username":"kid"}""", Json.EncodeBodyToString(new UpdateUserRequest("kid")));

    [Fact]
    public void QuotaLimitRoundTrips()
    {
        Assert.Equal("null", Json.EncodeBodyToString(QuotaLimit.None));
        Assert.Equal("12", Json.EncodeBodyToString(new QuotaLimit(12)));
    }
}

File: mac/MarqueeTests/RequestLimitsTests.swift
import XCTest
@testable import Marquee

// Request limits and the trusted role (0.39+): `/me` `requestLimits`, the
// quota fields on household members, the `PATCH /users/{id}` body, and who
// gets the review UI — decoded from the doc fixtures and from an older server.

final class RequestLimitsTests: XCTestCase {
    private func fixtureData(_ name: String) throws -> Data {
        try Data(contentsOf: Bundle(for: Self.self).resourceURL!.appendingPathComponent("Fixtures/api/\(name).json"))
    }

    private func decode<T: Decodable>(_ type: T.Type, _ name: String) throws -> T {
        try APIClient.decoder.decode(type, from: fixtureData(name))
    }

    private func decode<T: Decodable>(_ type: T.Type, json: String) throws -> T {
        try APIClient.decoder.decode(type, from: Data(json.utf8))
    }

    private func encodedObject(_ request: API.UpdateUserRequest) throws -> [String: Any] {
        try XCTUnwrap(JSONSerialization.jsonObject(with: APIClient.encoder.encode(request)) as? [String: Any])
    }

    private static let utc: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        return calendar
    }()

    private static let meBase = """
    "id":"54caac33-73d6-4864-8e12-1ea6b212d2f1","username":"kid","displayName":null,"role":"member",\
    "libraryOwnerId":"54caac33-73d6-4864-8e12-1ea6b212d2f1","autoApproveMovies":false,"autoApproveTv":false,\
    "createdAt":"2026-09-17T17:10:57.821Z"
    """

    private static let memberBase = """
    "id":"83c55a49-6153-4cb9-ae22-4a42d48f4cf3","username":"kid","displayName":null,\
    "autoApproveMovies":false,"autoApproveTv":false,"createdAt":"2026-09-17T17:12:40.991Z","isCurrentUser":false
    """

    // MARK: Decoding

    func testMeFixtureDecodesRequestLimits() throws {
        let me = try decode(API.Me.self, "me")
        let limits = try XCTUnwrap(me.requestLimits)
        XCTAssertNil(limits.movie)
        XCTAssertNil(limits.tv)
        XCTAssertNil(limits.summaryLine(), "No line when nothing is limited")
        XCTAssertTrue(me.canReviewRequests)
    }

    func testMeWithLimitsDecodes() throws {
        let json = """
        {\(Self.meBase),"requestLimits":{\
        "movie":{"limit":5,"days":7,"used":2,"remaining":3,"nextSlotAt":null},\
        "tv":{"limit":2,"days":30,"used":2,"remaining":0,"nextSlotAt":"2026-10-03T02:53:36.305Z"}}}
        """
        let me = try decode(API.Me.self, json: json)
        let limits = try XCTUnwrap(me.requestLimits)
        XCTAssertEqual(limits.movie, API.RequestLimit(limit: 5, days: 7, used: 2, remaining: 3, nextSlotAt: nil))
        XCTAssertEqual(limits.tv?.remaining, 0)
        XCTAssertNotNil(limits.tv?.nextSlotAt)
        XCTAssertEqual(
            limits.summaryLine(calendar: Self.utc),
            "Movies: 3 of 5 requests left (every 7 days) · TV: none left until Oct 3"
        )
        XCTAssertFalse(me.canReviewRequests)
    }

    func testQuotaLines() {
        let movie = API.RequestLimit(limit: 5, days: 7, used: 5, remaining: 0, nextSlotAt: nil)
        XCTAssertEqual(movie.line(label: "Movies"), "Movies: none left")
        XCTAssertEqual(
            API.RequestLimits(tv: API.RequestLimit(limit: 3, days: 14, used: 0, remaining: 3)).summaryLine(),
            "TV: 3 of 3 requests left (every 14 days)"
        )
    }

    func testHouseholdFixturesDecodeQuotaFields() throws {
        let member = try decode(API.HouseholdMember.self, "household-member")
        XCTAssertEqual(member.movieQuotaLimit, 5)
        XCTAssertEqual(member.movieQuotaDays, 7)
        XCTAssertNil(member.tvQuotaLimit)
        XCTAssertEqual(member.tvQuotaDays, 7)
        XCTAssertTrue(member.reportsRequestLimits)
        XCTAssertFalse(member.isTrusted)

        let update = try decode(API.UpdateUserResult.self, "user-update")
        XCTAssertEqual(update.user.movieQuotaLimit, 5)
        let users = try decode(API.ListResponse<API.HouseholdMember>.self, "users")
        XCTAssertFalse(users.results.isEmpty)
        XCTAssertTrue(users.results.allSatisfy(\.reportsRequestLimits))
        let imported = try decode(API.ImportUsersResult.self, "users-import-result")
        XCTAssertEqual(imported.created.first?.movieQuotaLimit, 5)
    }

    func testTrustedMemberDecodes() throws {
        let json = """
        {\(Self.memberBase),"role":"trusted","movieQuotaLimit":null,"movieQuotaDays":7,"tvQuotaLimit":null,"tvQuotaDays":7}
        """
        let member = try decode(API.HouseholdMember.self, json: json)
        XCTAssertEqual(member.role, .trusted)
        XCTAssertTrue(member.role.isKnown)
        XCTAssertTrue(member.isTrusted)
        XCTAssertFalse(member.isAdmin)
        XCTAssertEqual(member.role.label, "Trusted")
    }

    func testOlderServerDecodesWithoutTheNewFields() throws {
        let me = try decode(API.Me.self, json: "{\(Self.meBase)}")
        XCTAssertNil(me.requestLimits)

        let member = try decode(API.HouseholdMember.self, json: "{\(Self.memberBase),\"role\":\"member\"}")
        XCTAssertNil(member.movieQuotaLimit)
        XCTAssertNil(member.movieQuotaDays)
        XCTAssertFalse(member.reportsRequestLimits, "No role or limit controls for a server without them")
    }

    // MARK: Role gating

    func testOnlyAdminAndTrustedReviewRequests() {
        XCTAssertTrue(API.UserRole.admin.canReviewRequests)
        XCTAssertTrue(API.UserRole.trusted.canReviewRequests)
        XCTAssertFalse(API.UserRole.member.canReviewRequests)
        XCTAssertFalse(API.UserRole(rawValue: "guest").canReviewRequests, "An unknown role acts as a member")
        XCTAssertEqual(API.UserRole(rawValue: "trusted"), .trusted)

        let user = User(id: UUID(), username: "kid", displayName: nil, role: .trusted, libraryOwnerId: UUID())
        XCTAssertTrue(user.canReviewRequests)
        XCTAssertFalse(user.isAdmin, "Trusted doesn't get settings, integrations or Add buttons")
    }

    // MARK: PATCH body

    func testUpdateBodyOmitsUntouchedFields() throws {
        let body = try encodedObject(API.UpdateUserRequest(username: "kid", autoApproveTv: true))
        XCTAssertEqual(Set(body.keys), ["username", "autoApproveTv"])
    }

    func testUpdateBodySendsRoleAndLimits() throws {
        let request = API.UpdateUserRequest(
            username: "kid",
            role: .member,
            movieQuota: API.UpdateUserRequest.QuotaChange(limitText: "5", daysText: "7"),
            tvQuota: API.UpdateUserRequest.QuotaChange(limitText: "", daysText: "30")
        )
        let body = try encodedObject(request)
        XCTAssertEqual(body["role"] as? String, "member")
        XCTAssertEqual(body["movieQuotaLimit"] as? Int, 5)
        XCTAssertEqual(body["movieQuotaDays"] as? Int, 7)
        XCTAssertTrue(body["tvQuotaLimit"] is NSNull, "A blank limit is sent as null (removes it)")
        XCTAssertEqual(body["tvQuotaDays"] as? Int, 30)
    }

    func testQuotaFieldsParse() {
        XCTAssertEqual(API.UpdateUserRequest.QuotaChange(limitText: " 3 ", daysText: "7"), .init(limit: 3, days: 7))
        XCTAssertEqual(API.UpdateUserRequest.QuotaChange(limitText: "", daysText: "7"), .init(limit: nil, days: 7))
        XCTAssertNil(API.UpdateUserRequest.QuotaChange(limitText: "five", daysText: "7"))
        XCTAssertNil(API.UpdateUserRequest.QuotaChange(limitText: "5", daysText: ""))
    }
}
